pkg/state/impl/etcd/etcd.go

```diff
@@ -0,0 +1,578 @@
+func (st *State) Watch(ctx context.Context, resourcePointer resource.Pointer, ch chan<- state.Event, opts ...state.WatchOption) error {
+	var options state.WatchOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	if options.TailEvents > 0 {
+		return ErrUnsupported("tailEvents")
+	}
+
+	etcdKey := etcdKeyFromPointer(resourcePointer)
+
+	var (
+		curResource resource.Resource
+		err         error
+	)
+
+	getResp, err := st.cli.Get(ctx, etcdKey)
+	if err != nil {
+		return err
+	}
+
+	var initialEvent state.Event
+
+	if len(getResp.Kvs) > 0 {
+		curResource, err = st.unmarshalResource(getResp.Kvs[0])
+		if err != nil {
+			return err
+		}
+
+		initialEvent.Resource = curResource
+		initialEvent.Type = state.Created
+	} else {
+		initialEvent.Resource = resource.NewTombstone(
+			resource.NewMetadata(
+				resourcePointer.Namespace(),
+				resourcePointer.Type(),
+				resourcePointer.ID(),
+				resource.VersionUndefined,
+			))
+		initialEvent.Type = state.Destroyed
+	}
+
+	revision := getResp.Header.Revision
+
+	watchCh := st.cli.Watch(ctx, etcdKey, clientv3.WithPrevKV(), clientv3.WithRev(revision))
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
+		case ch <- initialEvent:
+		}
+
+		for {
+			var watchResponse clientv3.WatchResponse
+
+			select {
+			case <-ctx.Done():
+				return
+			case watchResponse = <-watchCh:
+			}
+
+			if watchResponse.Canceled {
+				return
+			}
+
+			for _, etcdEvent := range watchResponse.Events {
+				event, err := st.convertEvent(etcdEvent)
+				if err != nil {
+					// skip the event
+					continue
+				}
+
+				select {
+				case <-ctx.Done():
+					return
+				case ch <- *event:
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+// WatchKind all resources by type.
+//
+//nolint:gocyclo,cyclop,gocognit
+func (st *State) WatchKind(ctx context.Context, resourceKind resource.Kind, ch chan<- state.Event, opts ...state.WatchKindOption) error {
+	var options state.WatchKindOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	matches := func(res resource.Resource) bool {
+		return options.LabelQuery.Matches(*res.Metadata().Labels())
+	}
+
+	if options.TailEvents > 0 {
+		return ErrUnsupported("tailEvents")
+	}
+
+	etcdKey := etcdKeyPrefixFromKind(resourceKind)
+
+	var bootstrapList []resource.Resource
+
+	var revision int64
+
+	if options.BootstrapContents {
+		getResp, err := st.cli.Get(ctx, etcdKey, clientv3.WithPrefix())
+		if err != nil {
+			return err
+		}
+
+		revision = getResp.Header.Revision
+
+		for _, kv := range getResp.Kvs {
+			res, err := st.unmarshalResource(kv)
+			if err != nil {
+				return err
+			}
+
+			if !matches(res) {
+				continue
+			}
+
+			bootstrapList = append(bootstrapList, res)
+		}
+
+		sort.Slice(bootstrapList, func(i, j int) bool {
+			return bootstrapList[i].Metadata().ID() < bootstrapList[j].Metadata().ID()
+		})
+	}
+
+	watchCh := st.cli.Watch(ctx, etcdKey, clientv3.WithPrefix(), clientv3.WithPrevKV(), clientv3.WithRev(revision))
```

Comment:
if we don't set the `revision` (it's zero), `clientv3.WithRev(0)` means "current revision"?

Reply:
Yes, that seems to be the case.